workflow/torch/datastream.py

```diff
@@ -1,84 +1,343 @@
 from functools import partial
 from itertools import repeat, chain, islice
+from collections import namedtuple
 import numpy as np
 import pandas as pd
 import torch
 from workflow.functional import starcompose, star, repeat_map_chain
 from workflow.torch.dataset import Dataset
 
 
-class BoundlessSampler(torch.utils.data.Sampler):
-    def __init__(self, fn, length):
+class StandardSampler(torch.utils.data.WeightedRandomSampler):
+    def __init__(self, length, proportion=1.0, replacement=False):
+        super().__init__(
+            torch.ones(length).double(),
+            num_samples=int(length * proportion),
+            replacement=replacement,
+        )
+
+    def weight(self, index):
+        return self.weights[index].item()
+
+    def update_weights_(self, function):
+        self.weights[:] = function(self.weights)
+        
+    def update_example_weight_(self, weight, index):
+        if hasattr(weight, 'item'):
+            weight = weight.item()
+
+        self.weights[index] = weight
+
+    def sample_proportion(self, proportion):
+        sampler = StandardSampler(
+            len(self),
+            proportion,
+            self.replacement,
+        )
+        sampler.weights = self.weights
+        return sampler
+
+    def state_dict(self):
+        return dict(weights=self.weights)
+
+    def load_state_dict(self, state_dict):
+        self.weights[:] = state_dict['weights']
+
+
+class MergeSampler(torch.utils.data.Sampler):
+    def __init__(self, samplers, datasets, ns):
+        self.samplers = samplers
+        self.datasets = datasets
+        self.ns = ns
+        self.from_mapping = Dataset.create_from_concat_mapping(datasets)
+        self.merged_samplers = MergeSampler.merge_samplers(
+            samplers, datasets, ns
+        )
+        self.length = MergeSampler.merged_samplers_length(samplers)
+
+    def __len__(self):
+        return self.length
+
+    def __iter__(self):
+        return iter(self.merged_samplers)
+
+    @staticmethod
+    def merged_samplers_length(samplers):
+        # TODO: this is probably incorrect?
```

Comment:
when should we deal with this?

Reply:
The behavior of this function has been discussed a lot, there are arguments for all kinds of variants but we settled on this because it has the most practical convenience. We can remove the `TODO` comment I think.